rust_src/src/fns.rs

```diff
@@ -362,4 +362,32 @@ pub fn load_average(use_floats: bool) -> Vec<LispNumber> {
         .collect()
 }
 
+/// Return a copy of ALIST.
+/// This is an alist which represents the same mapping from objects to objects,
+/// but does not share the alist structure with ALIST.
+/// The objects mapped (cars and cdrs of elements of the alist)
+/// are shared, however.
+/// Elements of ALIST that are not conses are also shared.
+#[lisp_fn]
+pub fn copy_alist(mut alist: LispObject) -> LispObject {
+    if alist.is_nil() {
+        return alist;
+    }
+
+    let new_alist = unsafe { lisp_concat(1, &mut alist, Lisp_Type::Lisp_Cons, false) };
+
+    let mut tem = new_alist;
+    while tem.is_not_nil() {
```

Comment:
This looks like it could be written:

    for elt in tem.iter_tails(/* no checks */) {
        let front = elt.car();
        if let Some((car, cdr)) = front.into() {
            elt.set_car((car, cdr).into());
        }
    }

You are correct in asking if this could be a fold. If we had better control of allocations the pseudo code would look more like:

    let new_alist_vec = alist.iter_cars(/* no checks */).fold(Vec<LispObject>::new(), |accum, elt| {
        let new_cons: (LispObject, LispObject) = elt.into();
        accum.append(new_cons);
        accum
    });
    new_alist_vec.into()

Reply:
@nickdrozd I think this example shows a place where the casting style is less effective. How would you write this?

Reply:
What we want is a way to say:

    if elt.is_cons() {
        accum.append(copy of elt);
    }

Reply:
Would a `while let` simplify it?

Reply:
@nickdrozd I was referring to my hypothetical `fold` version. For now my first version or something close to it is probably the best.

```
for elt in tem.iter_tails(/* no checks */) {
    let front = elt.car();
    if let Some((car, cdr)) = front.into() {
        elt.set_car((car, cdr).into());
    }
}
```